test/e2e/framework/framework.go

```
@@ -488,6 +497,63 @@ func SendUDP(address, msg string) (string, error) {
 	return string(b[:n]), nil
 }
 
+// SendGameServerTCP sends a message to a gameserver and returns its reply
+// finds the first tcp port from the spec to send the message to,
+// returns error if no Ports were allocated
+func SendGameServerTCP(gs *agonesv1.GameServer, msg string) (string, error) {
+	if len(gs.Status.Ports) == 0 {
+		return "", errors.New("Empty Ports array")
+	}
+
+	// use first tcp port
+	for _, p := range gs.Spec.Ports {
+		if p.Protocol == corev1.ProtocolTCP {
+			return SendGameServerTCPToPort(gs, p.Name, msg)
+		}
+	}
+	return "", errors.New("No UDP ports")
+}
+
+// SendGameServerTCPToPort sends a message to a gameserver at the named port and returns its reply
+// returns error if no Ports were allocated or a port of the specified name doesn't exist
+func SendGameServerTCPToPort(gs *agonesv1.GameServer, portName string, msg string) (string, error) {
+	if len(gs.Status.Ports) == 0 {
+		return "", errors.New("Empty Ports array")
+	}
+	var port agonesv1.GameServerStatusPort
+	for _, p := range gs.Status.Ports {
+		if p.Name == portName {
+			port = p
+		}
+	}
+	address := fmt.Sprintf("%s:%d", gs.Status.Address, port.Port)
+	return SendTCP(address, msg)
+}
+
+// SendTCP sends a message to an address, and returns its reply if
+// it returns one in 30 seconds
+func SendTCP(address, msg string) (string, error) {
+	conn, err := net.Dial("tcp", address)
+	if err != nil {
+		log.Fatalf("Could not Dial: %v", err)
+	}
+
+	err = conn.SetReadDeadline(time.Now().Add(30 * time.Second))
+	if err != nil {
+		log.Fatalf("Timed out: %v", err)
+	}
+
+	defer func() {
+		err = conn.Close()
+	}()
+
+	// writes to the tcp connection
+	fmt.Fprintf(conn, msg+"\n")
+
+	response, _ := bufio.NewReader(conn).ReadString('\n')
```

This is ignoring the error from `ReadString`. Please handle the error case.